MoreLinq.NoConflictGenerator/Program.cs

```diff
@@ -0,0 +1,455 @@
+            // method with generic type parameters `TSource` and `TResult` will
+            // become `a` and `b`. This is used later for sorting and stabilizes
+            // the sort irrespective of how the type parameters are named or
+            // renamed in the source.
+            //
+
+            var abbreviatedTypeNodes = Enumerable
+                .Range(0, 26)
+                .Select(a => (char) ('a' + a))
+                .Select(ch => new SimpleTypeKey(ch.ToString()))
+                .ToArray();
+
+            var qq =
+                from fp in Directory.EnumerateFiles(dir, "*.cs")
+                where !excludePredicate(fp) && includePredicate(fp)
+                orderby fp
+                from m in
+                    from cd in
+                        CSharpSyntaxTree
+                            .ParseText(File.ReadAllText(fp), CSharpParseOptions.Default.WithPreprocessorSymbols("MORELINQ"))
+                            .GetRoot()
+                            .SyntaxTree
+                            .GetCompilationUnitRoot()
+                            .DescendantNodes().OfType<ClassDeclarationSyntax>()
+                    where (string) cd.Identifier.Value == "MoreEnumerable"
+                    from md in cd.DescendantNodes().OfType<MethodDeclarationSyntax>()
+                    let mn = (string) md.Identifier.Value
+                    where md.ParameterList.Parameters.Count > 0
+                       && md.ParameterList.Parameters.First().Modifiers.Any(m => (string)m.Value == "this")
+                       && md.Modifiers.Any(m => (string)m.Value == "public")
+                       && md.AttributeLists.SelectMany(al => al.Attributes).All(a => a.Name.ToString() != "Obsolete")
+                    let typeParameterAbbreviationByName =
+                        md.TypeParameterList
+                         ?.Parameters
+                          .Select((e, i) => (Original: e.Identifier.ValueText, Alias: abbreviatedTypeNodes[i]))
+                          .ToDictionary(e => e.Original, e => e.Alias)
+                    select new
+                    {
+                        Syntax = md,
+                        Name = md.Identifier.ToString(),
+                        TypeParameterCount = md.TypeParameterList?.Parameters.Count ?? 0,
+                        TypeParameterAbbreviationByName = typeParameterAbbreviationByName,
+                        ParameterCount = md.ParameterList.Parameters.Count,
+                        SortableParameterTypes =
+                            from p in md.ParameterList.Parameters
+                            select CreateTypeKey(p.Type,
+                                                 n => typeParameterAbbreviationByName != null
+                                                   && typeParameterAbbreviationByName.TryGetValue(n, out var a) ? a : null),
+                    }
+                select m;
```

Comment:
I'm not convinced. It adds back type ceremony that has no value. If we need clarification, I bet it can be done by:

- adding more vertical line-space to break the logical collections
- use `let` bindings to name the sub-parts/collections/queries
- add more comments

PS Those local functions wouldn't have been a possibility if we had been dealing with anonymous types.

Reply:
> I'm not convinced. It adds back type ceremony that has no value. If we need clarification, I bet it can be done by:
>
> - adding more vertical line-space to break the logical collections
> - use let bindings to name the sub-parts/collections/queries
> - add more comments

I don't think any of these options address the real issue, which is that currently I need to hold the entirety of the query in my head, as opposed to the split version, which requires holding a much smaller query in my head. I will need to internally split the query into subsections anyway, so why not have it be split in the source code and save me some work?

I agree the type ceremony is not particularly valuable, but I think the benefit outweighs the cost.

> PS Those local functions wouldn't have been a possibility if we had been dealing with anonymous types.

Right. Fortunately, this was not the case here.

Reply:
Your comment just made me realise that I hadn't pushed out the latest commits on the branch where I've attempted to break down the long query with comments. Have a look now and let me know if you think it's better than before.